### PR TITLE
Fix for ppc64le jit graph difference in sigmoid backward, see #10726

### LINKED ISSUE #10726: [ppc64le/pytorch] test_milstm_fusion_cuda and  test_lstm_fusion_cuda fails
The test using jit compiler seems ok for test_lstm_fusion_cuda  .  But not sure why the test_lstm_fusion_cuda (__main__.TestScript) fails. 

test_milstm_fusion_cuda fails (it has no equivalent jit test)
  
How can I correct this ?

Pytorch version :  torch (0.5.0a0+f0d8a36)
Cuda 9.2 
cudnn 7
ppc64le Power8

Here is the build/test log:  https://powerci.osuosl.org/job/pytorch-linux-cuda9-cudnn7-py3-mpi-build-test-gpu/264/console

``` 
======================================================================
FAIL: test_lstm_fusion_cuda (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/pytorch/test/common.py", line 251, in wrapper
    method(*args, **kwargs)
  File "/home/jenkins/pytorch/test/common.py", line 107, in wrapper
    fn(*args, **kwargs)
  File "test_jit.py", line 2591, in test_lstm_fusion_cuda
    self.assertExpectedGraph(backward_graph(module), subname='backward')
  File "test_jit.py", line 228, in assertExpectedGraph
    self.assertExpected(str(graph), *args, **kwargs)
  File "/home/jenkins/pytorch/test/common.py", line 514, in assertExpected
    self.assertMultiLineEqual(expected, s)
AssertionError: 'grap[908 chars]aph(%9 : Float(3, 20)\n      %19 : Float(3, 20[1722 chars]n}\n' != 'grap[908 chars]aph(%11 : Float(3, 20)\n      %21 : Float(3, 2[1723 chars]n}\n'
  graph(%0 : Float(3, 20!)
        %1 : Float(3, 20!)
        %2 : Dynamic
        %3 : Dynamic
        %4 : Dynamic
        %5 : Dynamic
        %6 : Dynamic
        %7 : Dynamic
        %8 : Dynamic
        %x : Float(3, 10)
        %hx : Float(3, 20)
        %cx : Float(3, 20)
        %12 : Float(10!, 80!)
        %13 : Float(20!, 80!)
        %ingate : Float(3, 20)
        %forgetgate : Float(3, 20)
        %cellgate : Float(3, 20)
        %outgate : Float(3, 20)
        %18 : Float(3, 20)) {
    %19 : int = prim::Constant[value=1]()
    %20 : Float(3, 80) = prim::FusionGroup_0[device=0](%ingate, %forgetgate, %cellgate, %outgate, %cx, %1, %18, %0)
    %21 : Float(3, 80) = aten::mul(%20, %19)
    %22 : Float(80!, 3!) = aten::t(%21)
    %23 : Float(80, 20) = aten::mm(%22, %hx)
    %24 : Float(80!, 3!) = aten::t(%20)
    %25 : Float(80, 10) = aten::mm(%24, %x)
    return (%25, %23, %21, %21);
  }
- with prim::FusionGroup_0 = graph(%9 : Float(3, 20)
?                                   ^
+ with prim::FusionGroup_0 = graph(%11 : Float(3, 20)
?                                   ^^
-       %19 : Float(3, 20)
?         -
+       %21 : Float(3, 20)
?        +
        %33 : Float(3, 20)
-       %39 : Float(3, 20)
?        ^^
+       %41 : Float(3, 20)
?        ^^
        %46 : Float(3, 20)
        %53 : Float(3, 20!)
        %65 : Float(3, 20)
        %67 : Float(3, 20!)) {
    %69 : Float(3, 20) = aten::mul(%67, %65)
-   %68 : Float(3, 20) = aten::mul(%67, %39)
?                                        ^^
+   %68 : Float(3, 20) = aten::mul(%67, %41)
?                                        ^^
    %66 : Float(3, 20) = aten::mul(%65, %65)
    %64 : Float(3, 20) = aten::neg(%66)
    %61 : int = prim::Constant[value=1]()
    %62 : Float(3, 20) = aten::add(%64, %61, %61)
    %59 : Float(3, 20) = aten::mul(%68, %62)
    %55 : int = prim::Constant[value=1]()
    %56 : Float(3, 20) = aten::add(%53, %59, %55)
    %51 : int = prim::Constant[value=1]()
    %52 : Float(3, 20) = aten::mul(%56, %51)
    %50 : Float(3, 20) = aten::mul(%52, %33)
-   %49 : Float(3, 20) = aten::mul(%52, %9)
?                                        ^
+   %49 : Float(3, 20) = aten::mul(%52, %11)
?                                        ^^
    %47 : Float(3, 20) = aten::mul(%56, %46)
+   %44 : Float(3, 20) = aten::mul(%69, %41)
-   %44 : Float(3, 20) = aten::neg(%39)
?     ^                             ^^
+   %42 : Float(3, 20) = aten::neg(%41)
?     ^                             ^^
-   %42 : int = prim::Constant[value=1]()
?    ^^
+   %39 : int = prim::Constant[value=1]()
?    ^^
-   %43 : Float(3, 20) = aten::add(%44, %42, %42)
-   %40 : Float(3, 20) = aten::mul(%69, %39)
?                              ^^^  ^
+   %40 : Float(3, 20) = aten::add(%42, %39, %39)
?                              ^^^  ^^^^^^
-   %37 : Float(3, 20) = aten::mul(%40, %43)
?                                    ^    ^
+   %37 : Float(3, 20) = aten::mul(%44, %40)
?                                    ^    ^
    %34 : Float(3, 20) = aten::mul(%33, %33)
    %32 : Float(3, 20) = aten::neg(%34)
    %29 : int = prim::Constant[value=1]()
    %30 : Float(3, 20) = aten::add(%32, %29, %29)
    %27 : Float(3, 20) = aten::mul(%49, %30)
-   %24 : Float(3, 20) = aten::neg(%19)
-   %22 : int = prim::Constant[value=1]()
-   %23 : Float(3, 20) = aten::add(%24, %22, %22)
-   %20 : Float(3, 20) = aten::mul(%47, %19)
?     ^                                   -
+   %24 : Float(3, 20) = aten::mul(%47, %21)
?     ^                                  +
-   %17 : Float(3, 20) = aten::mul(%20, %23)
-   %14 : Float(3, 20) = aten::neg(%9)
?    ^^                             ^
+   %22 : Float(3, 20) = aten::neg(%21)
?    ^^                             ^^
-   %12 : int = prim::Constant[value=1]()
?     ^
+   %19 : int = prim::Constant[value=1]()
?     ^
-   %13 : Float(3, 20) = aten::add(%14, %12, %12)
?    ^^                             ^^    ^    ^
+   %20 : Float(3, 20) = aten::add(%22, %19, %19)
?    ^^                             ^^    ^    ^
+   %17 : Float(3, 20) = aten::mul(%24, %20)
-   %10 : Float(3, 20) = aten::mul(%50, %9)
?     ^                                  ^
+   %14 : Float(3, 20) = aten::mul(%50, %11)
?     ^                                  ^^
+   %12 : Float(3, 20) = aten::neg(%11)
+   %9 : int = prim::Constant[value=1]()
+   %10 : Float(3, 20) = aten::add(%12, %9, %9)
-   %7 : Float(3, 20) = aten::mul(%10, %13)
?                                   ^    ^
+   %7 : Float(3, 20) = aten::mul(%14, %10)
?                                   ^    ^
    %4 : Float(3, 80) = prim::FusedConcat[dim=1](%7, %17, %27, %37)
    return (%4);
  }


======================================================================
FAIL: test_milstm_fusion_cuda (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/pytorch/test/common.py", line 251, in wrapper
    method(*args, **kwargs)
  File "/home/jenkins/pytorch/test/common.py", line 107, in wrapper
    fn(*args, **kwargs)
  File "test_jit.py", line 2603, in test_milstm_fusion_cuda
    self.assertExpectedGraph(backward_graph(module), subname='backward')
  File "test_jit.py", line 228, in assertExpectedGraph
    self.assertExpected(str(graph), *args, **kwargs)
  File "/home/jenkins/pytorch/test/common.py", line 514, in assertExpected
    self.assertMultiLineEqual(expected, s)
AssertionError: 'grap[1502 chars]aph(%9 : Float(3, 20)\n      %19 : Float(3, 20[2751 chars]n}\n' != 'grap[1502 chars]aph(%11 : Float(3, 20)\n      %21 : Float(3, 2[2752 chars]n}\n'
  graph(%0 : Float(3, 20!)
        %1 : Float(3, 20!)
        %2 : Dynamic
        %3 : Dynamic
        %4 : Dynamic
        %5 : Dynamic
        %6 : Dynamic
        %7 : Dynamic
        %8 : Dynamic
        %9 : Dynamic
        %10 : Dynamic
        %11 : Dynamic
        %x : Float(3, 10)
        %hx : Float(3, 20)
        %cx : Float(3, 20)
        %alpha : Float(80)
        %beta_i : Float(80)
        %beta_h : Float(80)
        %18 : Float(10!, 80!)
        %Wx : Float(3, 80)
        %20 : Float(20!, 80!)
        %Uz : Float(3, 80)
        %22 : Float(3, 80)
        %ingate : Float(3, 20)
        %forgetgate : Float(3, 20)
        %cellgate : Float(3, 20)
        %outgate : Float(3, 20)
        %27 : Float(3, 20)) {
    %28 : Float(3, 80) = prim::FusionGroup_0[device=0](%ingate, %forgetgate, %cellgate, %outgate, %cx, %1, %27, %0)
    %29 : int[] = prim::Constant[value=[3, 80]]()
    %30 : int = prim::Constant[value=0]()
    %31 : Float(3!, 80) = aten::expand(%beta_h, %29, %30)
    %32 : Float(3!, 80) = aten::expand(%beta_i, %29, %30)
    %33 : Float(3!, 80) = aten::expand(%alpha, %29, %30)
    %34 : Float(3, 80), %35 : Float(3, 80), %36 : Float(3, 80), %37 : Float(3, 80), %38 : Float(3, 80), %39 : Float(3, 80) = prim::FusionGroup_1[device=0](%33, %32, %Wx, %28, %Uz, %22, %31)
    %40 : Float(80!, 3!) = aten::t(%36)
    %41 : Float(80, 20) = aten::mm(%40, %hx)
    %42 : Float(80!, 3!) = aten::t(%34)
    %43 : Float(80, 10) = aten::mm(%42, %x)
    return (%43, %41, %35, %37, %38, %39);
  }
- with prim::FusionGroup_0 = graph(%9 : Float(3, 20)
?                                   ^
+ with prim::FusionGroup_0 = graph(%11 : Float(3, 20)
?                                   ^^
-       %19 : Float(3, 20)
?         -
+       %21 : Float(3, 20)
?        +
        %33 : Float(3, 20)
-       %39 : Float(3, 20)
?        ^^
+       %41 : Float(3, 20)
?        ^^
        %46 : Float(3, 20)
        %53 : Float(3, 20!)
        %65 : Float(3, 20)
        %67 : Float(3, 20!)) {
    %69 : Float(3, 20) = aten::mul(%67, %65)
-   %68 : Float(3, 20) = aten::mul(%67, %39)
?                                        ^^
+   %68 : Float(3, 20) = aten::mul(%67, %41)
?                                        ^^
    %66 : Float(3, 20) = aten::mul(%65, %65)
    %64 : Float(3, 20) = aten::neg(%66)
    %61 : int = prim::Constant[value=1]()
    %62 : Float(3, 20) = aten::add(%64, %61, %61)
    %59 : Float(3, 20) = aten::mul(%68, %62)
    %55 : int = prim::Constant[value=1]()
    %56 : Float(3, 20) = aten::add(%53, %59, %55)
    %51 : int = prim::Constant[value=1]()
    %52 : Float(3, 20) = aten::mul(%56, %51)
    %50 : Float(3, 20) = aten::mul(%52, %33)
-   %49 : Float(3, 20) = aten::mul(%52, %9)
?                                        ^
+   %49 : Float(3, 20) = aten::mul(%52, %11)
?                                        ^^
    %47 : Float(3, 20) = aten::mul(%56, %46)
+   %44 : Float(3, 20) = aten::mul(%69, %41)
-   %44 : Float(3, 20) = aten::neg(%39)
?     ^                             ^^
+   %42 : Float(3, 20) = aten::neg(%41)
?     ^                             ^^
-   %42 : int = prim::Constant[value=1]()
?    ^^
+   %39 : int = prim::Constant[value=1]()
?    ^^
-   %43 : Float(3, 20) = aten::add(%44, %42, %42)
-   %40 : Float(3, 20) = aten::mul(%69, %39)
?                              ^^^  ^
+   %40 : Float(3, 20) = aten::add(%42, %39, %39)
?                              ^^^  ^^^^^^
-   %37 : Float(3, 20) = aten::mul(%40, %43)
?                                    ^    ^
+   %37 : Float(3, 20) = aten::mul(%44, %40)
?                                    ^    ^
    %34 : Float(3, 20) = aten::mul(%33, %33)
    %32 : Float(3, 20) = aten::neg(%34)
    %29 : int = prim::Constant[value=1]()
    %30 : Float(3, 20) = aten::add(%32, %29, %29)
    %27 : Float(3, 20) = aten::mul(%49, %30)
-   %24 : Float(3, 20) = aten::neg(%19)
-   %22 : int = prim::Constant[value=1]()
-   %23 : Float(3, 20) = aten::add(%24, %22, %22)
-   %20 : Float(3, 20) = aten::mul(%47, %19)
?     ^                                   -
+   %24 : Float(3, 20) = aten::mul(%47, %21)
?     ^                                  +
-   %17 : Float(3, 20) = aten::mul(%20, %23)
-   %14 : Float(3, 20) = aten::neg(%9)
?    ^^                             ^
+   %22 : Float(3, 20) = aten::neg(%21)
?    ^^                             ^^
-   %12 : int = prim::Constant[value=1]()
?     ^
+   %19 : int = prim::Constant[value=1]()
?     ^
-   %13 : Float(3, 20) = aten::add(%14, %12, %12)
?    ^^                             ^^    ^    ^
+   %20 : Float(3, 20) = aten::add(%22, %19, %19)
?    ^^                             ^^    ^    ^
+   %17 : Float(3, 20) = aten::mul(%24, %20)
-   %10 : Float(3, 20) = aten::mul(%50, %9)
?     ^                                  ^
+   %14 : Float(3, 20) = aten::mul(%50, %11)
?     ^                                  ^^
+   %12 : Float(3, 20) = aten::neg(%11)
+   %9 : int = prim::Constant[value=1]()
+   %10 : Float(3, 20) = aten::add(%12, %9, %9)
-   %7 : Float(3, 20) = aten::mul(%10, %13)
?                                   ^    ^
+   %7 : Float(3, 20) = aten::mul(%14, %10)
?                                   ^    ^
    %4 : Float(3, 80) = prim::FusedConcat[dim=1](%7, %17, %27, %37)
    return (%4);
  }
  with prim::FusionGroup_1 = graph(%5 : Float(3!, 80)
        %8 : Float(3!, 80)
        %10 : Float(3, 80)
        %12 : Float(3, 80)
        %13 : Float(3, 80)
        %20 : Float(3, 80)
        %22 : Float(3!, 80)) {
    %30 : int = prim::Constant[value=1]()
    %29 : int = prim::Constant[value=1]()
    %28 : int = prim::Constant[value=1]()
    %26 : int = prim::Constant[value=1]()
    %27 : Float(3, 80) = aten::mul(%12, %26)
    %25 : Float(3, 80) = aten::mul(%27, %13)
    %24 : Float(3, 80) = aten::mul(%27, %10)
    %23 : Float(3, 80) = aten::mul(%27, %22)
    %21 : Float(3, 80) = aten::mul(%12, %20)
    %19 : int = prim::Constant[value=1]()
    %17 : int = prim::Constant[value=1]()
    %18 : Float(3, 80) = aten::add(%23, %21, %17)
    %14 : Float(3, 80) = aten::mul(%12, %13)
    %11 : Float(3, 80) = aten::mul(%14, %10)
    %9 : Float(3, 80) = aten::mul(%27, %8)
    %6 : Float(3, 80) = aten::mul(%14, %5)
    %2 : int = prim::Constant[value=1]()
    %3 : Float(3, 80) = aten::add(%9, %6, %2)
    return (%3, %11, %18, %24, %25, %27);
  }

### PR DESCRIPTION
**Pull request by avmgithub**
As reported in Issue #10726, the jit compiler, when running on ppc64le,  may produce an isomorphic output but fail a diff test against the expected output file.  The expected output file is created from a test that was ran on x86_64.  This ensures that if ppc64le test output is different, the output is instead compared to an expected output file created when the test is run on a ppc64le system.

### CODE REVIEW

**Review by ezyang — DISMISSED** [review]
This is not really sustainable; most devs will not be able to keep the ppc files up to date. It would be better to figure out why results are different on ppc.

**Comment by avmgithub** [comment]
@ezyang no problem, I understand.  Is there anyway you can give us some hints on where this problem maybe in the source.  We had a similar problem before reported in issue #5055. And it was fixed in #5124.  That was all in c++ code.  This one is in Python code.

**Comment by ezyang** [comment]
So, the first thing I would check is that the canonicalizer is working on subgraphs. There don't appear to be structural differences; and if that's the case, the graphs SHOULD number identically. It really looks like we're not canonicalizing subgraphs.

**Comment by avmgithub** [comment]
@ezyang are you referring to the Canonicalize function here https://github.com/pytorch/pytorch/blob/00df09b65d9f07bec7ee693658f017fda30a27e7/torch/csrc/jit/passes/canonicalize.cpp#L7    ?  And when you refer to "not canonicalizing subgraphs"  are you referring to :  https://github.com/pytorch/pytorch/blob/00df09b65d9f07bec7ee693658f017fda30a27e7/torch/csrc/jit/passes/canonicalize.cpp#L28   ?

It looks to me they are fine.  If they are not would I not see the problem also happen in the forward graph?

Just to add to the problem description:  The problem only happens for the backward graph .  The forward graph is fine.  

So for example part of the output of the backward graph of  the test_milstm_fusion_cuda is below: 

Output of ppc64le
%22 : Float(*, *) = aten::mul(%8, %3)
%23 : Float(*, *) = aten::neg(%3)
%24 : int = prim::Constant[value=1]()
%25 : Float(*, *) = aten::add(%23, %24, %24)
%26 : Float(*, *) = aten::mul(%22, %25)

Output of x86
%22 : Float(*, *) = aten::neg(%3)
%23 : int = prim::Constant[value=1]()
%24 : Float(*, *) = aten::add(%22, %23, %23)
%25 : Float(*, *) = aten::mul(%8, %3)
%26 : Float(*, *) = aten::mul(%25, %24)

So it seems the instructions are kind of re-arranged.    Do you think this is due to canonicalization ?
I'm thinking something to do with the backward functionality.  But I may be totally wrong.  Need your expert opinion and help as to which source file I need to be looking at.

**Comment by zou3519** [comment]
@avmgithub from not knowing very much about what is wrong, I would start from https://github.com/pytorch/pytorch/blob/e00fb69b251c7a1af2b172cc5c059945686b71b5/torch/csrc/jit/graph_executor.cpp#L372 and print out the graph (`std::cout << *graph << std::endl;`) for a ppc64le arch and x86 arch and see where it differs

**Comment by avmgithub** [comment]
@zou3519 , I've attached 2 files , one for x86 and one ppc64le .  From a test run like : python -m unittest -q  test_jit.TestScript.test_milstm_fusion_cuda   .     There's differences in the sigmoid function. I'm assuming the graph is for the backward operation.   Here is an example:

x86 : 
%56 : Dynamic = aten::neg(%outgate)
%57 : Dynamic = aten::add(%56, %28, %28)
%58 : Dynamic = aten::mul(%34, %outgate)
%59 : Dynamic = aten::mul(%58, %57)

ppc : 
%56 : Dynamic = aten::mul(%34, %outgate)
%57 : Dynamic = aten::neg(%outgate)
%58 : Dynamic = aten::add(%57, %28, %28)
%59 : Dynamic = aten::mul(%56, %58)

There are 3 such instances since there are three sigmoid functions in the MiLSTMCell  function in test_jit.py

**Comment by avmgithub** [comment]
@zou3519 When you get the chance, do know where (which source file)  the sigmoid backward graph is formed ?  This seems to be where the discrepancy is coming from.

**Comment by zou3519** [comment]
it depends, but it is most likely coming from autodiff here: https://github.com/pytorch/pytorch/blob/76ab26cc3eff1d7ba822d8db93723f5c9598eead/torch/csrc/jit/autodiff.cpp#L101

**Comment by avmgithub** [comment]
@zou3519 Thanks for the tip.  It looks like if I re-arrange the line : 
return {grads.at(0) * outputs.at(0) * (1 - outputs.at(0))};
to 
return {((1 - outputs.at(0)) * outputs.at(0) * grads.at(0))};

And re-run the :  python test/test_jit.py TestScript.test_milstm_fusion_cuda --accept 
both expect files (forward and backward)  are now identical.

Can you please let me know if this is OK to do.  

I have no idea why I have to re-arrange it to give me the same backward expect output.

**Comment by avmgithub** [comment]
@ezyang @zou3519 When you get the chance, please review suggested changes.

**Comment by sdmonov** [comment]
I tested the proposed solution on a ppc64le and it works for me.

File: torch/csrc/jit/autodiff.cpp [hunk 1]
```diff
@@ -99,7 +99,7 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
       return {grads.at(0) * inputs.at(1), grads.at(0) * inputs.at(0)};
 
     } else if (node->matches("aten::sigmoid(Tensor self) -> Tensor")) {
-      return {grads.at(0) * outputs.at(0) * (1 - outputs.at(0))};
+      return {(1 - outputs.at(0)) * outputs.at(0) * grads.at(0)};
```
Comment (zou3519) [inline torch/csrc/jit/autodiff.cpp]: Please add a comment/TODO here that the order of operations matters for different architectures and to look into it

**Review by zou3519 — APPROVED** [review]
It's not a complete fix, but it works for now and is simple enough. Please add a TODO/comment into the code to prevent regressions and so that someone in the future will take a deeper look at it

File: torch/csrc/jit/autodiff.cpp (hunk 1 above)
Comment (avmgithub) [reply torch/csrc/jit/autodiff.cpp]: @zou3519 thanks for checking.  I added a TODO comment.

**Comment by avmgithub** [comment]
@zou3519 Thanks for approving,  was there something else that needs to be done before this can be merged.  It still says merging is blocked

**Review by soumith — APPROVED** [review]

**Pull request closed without merging**